File: src/main/java/brightspark/runicmagic/util/Location.java
package brightspark.runicmagic.util;

import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.math.BlockPos;
import net.minecraftforge.common.util.INBTSerializable;

public class Location implements INBTSerializable<NBTTagCompound>
{
    private int dimension;
    private BlockPos position;

    public Location(int dimension, BlockPos position)
    {
        this.dimension = dimension;
        this.position = position;
    }

    public Location(NBTTagCompound nbt)
    {
        deserializeNBT(nbt);
    }

    public int getDimension()
    {
        return dimension;
    }

    public BlockPos getPosition()
    {
        return position;
    }

    @Override
    public NBTTagCompound serializeNBT()
    {
        NBTTagCompound nbt = new NBTTagCompound();
        nbt.setInteger("dim", dimension);
        nbt.setLong("pos", position.toLong());
        return nbt;
    }

    @Override
    public void deserializeNBT(NBTTagCompound nbt)
    {
        dimension = nbt.getInteger("dim");
        position = BlockPos.fromLong(nbt.getLong("pos"));
    }
}

File: src/main/kotlin/brightspark/runicmagic/model/Location.kt
package brightspark.runicmagic.model

import net.minecraft.nbt.CompoundNBT
import net.minecraft.util.RegistryKey
import net.minecraft.util.ResourceLocation
import net.minecraft.util.math.BlockPos
import net.minecraft.util.registry.Registry
import net.minecraft.world.World
import net.minecraftforge.common.util.INBTSerializable

class Location : INBTSerializable<CompoundNBT> {
	lateinit var dimension: ResourceLocation
		private set
	lateinit var position: BlockPos
		private set
	val dimensionKey: RegistryKey<World> by lazy { RegistryKey.getOrCreateKey(Registry.WORLD_KEY, dimension) }

	constructor(dimension: ResourceLocation, position: BlockPos) {
		this.dimension = dimension
		this.position = position
	}

	constructor(nbt: CompoundNBT) {
		deserializeNBT(nbt)
	}

	override fun serializeNBT(): CompoundNBT = CompoundNBT().apply {
		putString("dim", dimension.toString())
		putLong("pos", position.toLong())
	}

	override fun deserializeNBT(nbt: CompoundNBT) {
		dimension = ResourceLocation(nbt.getString("dim"))
		position = BlockPos.fromLong(nbt.getLong("pos"))
	}
}